module/finalizer/collection/finalizer_test.go

```diff
@@ -180,15 +179,13 @@ func TestFinalizer(t *testing.T) {
 			assertClusterBlocksIndexedByReferenceHeight(t, db, refBlock.Height, final.ID())
 
 			// block should be passed to provider
-			prov.AssertNumberOfCalls(t, "SubmitLocal", 1)
-			prov.AssertCalled(t, "SubmitLocal", &messages.SubmitCollectionGuarantee{
-				Guarantee: flow.CollectionGuarantee{
-					CollectionID:     block.Payload.Collection.ID(),
-					ReferenceBlockID: refBlock.ID(),
-					ChainID:          block.Header.ChainID,
-					SignerIndices:    block.Header.ParentVoterIndices,
-					Signature:        nil,
-				},
+			prov.AssertNumberOfCalls(t, "SubmitCollectionGuarantee", 1)
+			prov.AssertCalled(t, "SubmitCollectionGuarantee", &flow.CollectionGuarantee{
+				CollectionID:     block.Payload.Collection.ID(),
+				ReferenceBlockID: refBlock.ID(),
+				ChainID:          block.Header.ChainID,
+				SignerIndices:    block.Header.ParentVoterIndices,
+				Signature:        nil,
 			})
```

Comment:
We can similarly change this call to be part of the assertion at the beginning of the test. We would remove lines 183-189 and replace line 150 with:
```go
prov.On("SubmitCollectionGuarantee", &flow.CollectionGuarantee{
	CollectionID:     block.Payload.Collection.ID(),
	ReferenceBlockID: refBlock.ID(),
	ChainID:          block.Header.ChainID,
	SignerIndices:    block.Header.ParentVoterIndices,
	Signature:        nil,
}).Once()
```

This is again a sign of older code using old patterns (not a problem with your PR: just pointing out something that is already present). When you write new tests, you should generally write specific assertions before execution, rather than afterward. It is less code, and having the assertions before execution avoids the potential pitfall of using the test output itself as the expected value.

I'm pointing this out mainly as a pattern to avoid directly replicating in your own code. But I do think it would be a good exercise to update the assertions in this test file as described above, to get a sense for how the library works and when/how assertions are validated.

Reply:
I did this in https://github.com/onflow/flow-go/pull/6780/commits/d157f489b4bc68713b46269a90be534c99a7ee0c, and also made sure to see the error messages when the test fails because those assertions aren't met; definitely useful that it tells you when the number of expected calls doesn't match.

I did run into what I think is a VSCode issue: when the tests are run from VSCode and fail, the file that the error occurred in is given the wrong path (it's assumed to be in the "module/finalizer/collection/" directory, the same one as the test is in); when the tests are instead run on the command line via `go test`, it only reports the filename where the error occurred (not a full path).